osu.Game/Overlays/Profile/UserGraph.cs

```diff
@@ -0,0 +1,235 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Cursor;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osuTK;
+
+namespace osu.Game.Overlays.Profile
+{
+    public abstract class UserGraph<TKey, TValue> : Container, IHasCustomTooltip
+    {
+        protected const float FADE_DURATION = 150;
+
+        protected readonly RankChartLineGraph Graph;
+        protected KeyValuePair<TKey, TValue>[] Data;
+        protected int DataIndex;
+
+        protected UserGraph()
+        {
+            Add(Graph = new RankChartLineGraph
+            {
+                RelativeSizeAxes = Axes.Both,
+                Alpha = 0
+            });
+
+            Graph.OnBallMove += i => DataIndex = i;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours)
+        {
+            Graph.LineColour = colours.Yellow;
+        }
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            if (Data?.Length > 1)
+            {
+                Graph.UpdateBallPosition(e.MousePosition.X);
+                Graph.ShowBar();
+            }
+
+            return base.OnHover(e);
+        }
+
+        protected override bool OnMouseMove(MouseMoveEvent e)
+        {
+            if (Data?.Length > 1)
+                Graph.UpdateBallPosition(e.MousePosition.X);
+
+            return base.OnMouseMove(e);
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            if (Data?.Length > 1)
+                Graph.HideBar();
+
+            base.OnHoverLost(e);
+        }
+
+        public ITooltip GetCustomTooltip() => GetTooltip();
+
+        public object TooltipContent => GetTooltipContent();
+
+        protected abstract UserGraphTooltip GetTooltip();
+
+        protected abstract object GetTooltipContent();
+
+        protected class RankChartLineGraph : LineGraph
```

Comment:
Needs to be renamed since this is now in the abstract. Maybe `HoverableLineGraph`? `UserLineGraph`?